nifi-nar-bundles/nifi-graph-bundle/nifi-graph-processors/src/main/java/org/apache/nifi/processors/graph/ExecuteGraphQueryRecord.java

```diff
@@ -0,0 +1,289 @@
+        String recordScript = context.getProperty(SUBMISSION_SCRIPT)
+                .evaluateAttributeExpressions(input)
+                .getValue();
+
+        boolean failed = false;
+        boolean error = false;
+        final AtomicLong errors = new AtomicLong();
+        long delta = 0;
+        Map<String, String> attributes = new HashMap<>();
+        List<Map<String,Object>> graphResponses = new ArrayList<>();
+        try (InputStream is = session.read(input);
+             OutputStream os = session.write(output);
+             OutputStream graphOutputStream = session.write(graph);
+             RecordReader reader = recordReaderFactory.createRecordReader(attributes, is, -1l, getLogger());
+             RecordSetWriter writer = recordSetWriterFactory.createWriter(getLogger(), reader.getSchema(), os, input.getAttributes());
+        ) {
+            List<PropertyDescriptor> dynamic = context.getProperties()
+                    .keySet().stream()
+                    .filter(PropertyDescriptor::isDynamic)
+                    .collect(Collectors.toList());
+            Record record;
+
+            long start = System.currentTimeMillis();
+            long recordIndex = 0;
+            long batchStart = start;
+            writer.beginRecordSet();
+            while ((record = reader.nextRecord()) != null) {
+                Map<String, Object> dynamicPropertyMap = new HashMap<>();
+                for (PropertyDescriptor entry : dynamic) {
+                        if(!dynamicPropertyMap.containsKey(entry.getName())) {
+                            dynamicPropertyMap.put(entry.getName(), getRecordValue(context, entry, record, input));
+                        }
+                }
+                dynamicPropertyMap.putAll(input.getAttributes());
+                graphResponses.addAll(executeQuery(recordScript, dynamicPropertyMap));
+                recordIndex++;
+                if (recordIndex % 100 == 0) {
+                    long now = System.currentTimeMillis();
+                    batchStart = System.currentTimeMillis();
+                }
+            }
+            try {
+                String graphOutput = mapper.writerWithDefaultPrettyPrinter().writeValueAsString(graphResponses);
+                graphOutputStream.write(graphOutput.getBytes(StandardCharsets.UTF_8));
+                session.getProvenanceReporter().modifyContent(graph);
+            } catch ( Exception ex ) {
+                getLogger().error("", ex);
+                error = true;
+            }
+            writer.finishRecordSet();
+            long end = System.currentTimeMillis();
+            delta = (end - start) / 1000;
+            if (getLogger().isDebugEnabled()){
+                getLogger().debug(String.format("Took %s seconds.", delta));
+            }
+        } catch (Exception ex) {
+            getLogger().error("", ex);
+            failed = true;
+        } finally {
+            if (failed) {
+                session.remove(output);
```

Comment:
Removing graph output should happen here.

Reply:
Refactored and removed on failure.